File: src/main/java/sx/concurrent/ManualResetEvent.java
package sx.concurrent;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

/**
 * Manual reset mLatch
 * @author masc
 */
public class ManualResetEvent implements Event {
    private volatile CountDownLatch mLatch;
    private Object mSync = new Object();

    public ManualResetEvent(boolean signalled) {
        if (signalled) {
            mLatch = new CountDownLatch(0);
        } else {
            mLatch = new CountDownLatch(1);
        }
    }

    public void set() {
        mLatch.countDown();
    }

    public void reset() {
        synchronized (mSync) {
            if (mLatch.getCount() == 0) {
                mLatch = new CountDownLatch(1);
            }
        }
    }

    public void waitOne() throws InterruptedException {
        mLatch.await();
    }

    public boolean waitOne(int timeout, TimeUnit unit) throws InterruptedException {
        return mLatch.await(timeout, unit);
    }

    public boolean isSignalled() {
        return mLatch.getCount() == 0;
    }
}

File: src/main/java/sx/concurrent/ManualResetEvent.kt
package sx.concurrent

import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * Manual reset mLatch
 * @author masc
 */
class ManualResetEvent(signalled: Boolean) : Event {
    @Volatile private var latch: CountDownLatch
    private val sync = ReentrantLock()

    init {
        if (signalled) {
            latch = CountDownLatch(0)
        } else {
            latch = CountDownLatch(1)
        }
    }

    override fun set() {
        latch.countDown()
    }

    override fun reset() {
        this.sync.withLock {
            if (latch.count == 0L) {
                latch = CountDownLatch(1)
            }
        }
    }

    @Throws(InterruptedException::class)
    override fun waitOne() {
        latch.await()
    }

    @Throws(InterruptedException::class)
    override fun waitOne(timeout: Int, unit: TimeUnit): Boolean {
        return latch.await(timeout.toLong(), unit)
    }

    override val isSignalled: Boolean
        get() = (latch.count == 0L)

}